src/infinitempomatrix.jl

```diff
@@ -44,111 +44,212 @@ function InfiniteMPOMatrix(data::Vector{Matrix{ITensor}}, translator::Function)
-#  We ITensors.directsum() to join all the blocks into the final ITensor.
-#  This code should be 100% dense/blocks-sparse agnostic.
-#
-function cat_to_itensor(Hm::Matrix{ITensor}, inds_array)::Tuple{ITensor,Index,Index}
+function InfiniteMPOMatrix(data::CelledVector{Matrix{ITensor}}, m::Int64, n::Int64)
+  return InfiniteMPOMatrix(data, m, n, false)
+end
+
+function InfiniteMPOMatrix(data::CelledVector{Matrix{ITensor}})
+  return InfiniteMPOMatrix(data, 0, size(data)[1], false)
+end
+
+function ITensors.siteinds(A::InfiniteMPOMatrix)
+  return CelledVector(
+    [dag(only(filterinds(A[x][1, 1]; plev=0, tags="Site"))) for x in 1:nsites(A)],
+    translator(A),
+  )
+end
+
+function ITensors.splitblocks(H::InfiniteMPOMatrix)
+  N = nsites(H)
+  for j in 1:N
+    for n in 1:length(H)
+      H[j][n] = splitblocks(H[j][n])
+    end
+  end
+  return H
+end
+
+function find_all_links(Hm::Matrix{ITensor})
+  is = inds(Hm[1, 1]) #site inds
   lx, ly = size(Hm)
-  @assert lx == ly
-  #
-  #  Extract the sub diagonal
-  #
-  Ms = map(n -> Hm[lx - n + 1, ly - n], 1:(lx - 1)) #Get the sub diagonal into an array.
-  N = length(Ms)
-  @assert N == lx - 1
-  #
-  # Convert edge Ms to order 4 ITensors using the dummy index.
-  #
-  il0 = inds_array[1][1]
-  T = eltype(Ms[1])
-  Ms[1] *= onehot(T, il0 => 1)
-  Ms[N] *= onehot(T, dag(il0) => 1) #We don't need distinct index IDs for this.  directsum makes all new index anyway.
-  #
-  # Start direct sums to build the single ITensor.  Order is critical to get the desired form.
-  #
-  H = Hm[1, 1] * onehot(T, il0' => 1) * onehot(T, dag(il0) => 1) #Bootstrap with the I op in the top left of Hm
-  H, il = directsum(H => il0', Ms[N] => inds_array[N][1]) #1-D directsum to put M[N] directly below the I op
-  ir = dag(il0) #setup recusion.
-  for i in (N - 1):-1:1 #2-D directsums to place new blocks below and to the right.
-    H, (il, ir) = directsum(
-      H => (il, ir), Ms[i] => inds_array[i]; tags=["Link,left", "Link,right"]
+  #We extract the links from the order-3 tensors on the first column and line
+  #We add dummy indices if there is no relevant indices
+  ir = only(uniqueinds(Hm[1, 2], is))
+  ir0 = Index(ITensors.trivial_space(ir); dir=dir(ir), tags="Link,extra")
+  il0 = dag(ir0)
+  left_links = typeof(ir)[]
+  for x in 1:lx
+    temp = uniqueinds(Hm[x, 1], is)
+    if length(temp) == 0
+      append!(left_links, [il0])
+    elseif length(temp) == 1
+      append!(left_links, temp)
+    else
+      error("ITensor does not seem to be of the correct order")
+    end
+  end
+  right_links = typeof(ir)[]
+  for x in 1:lx
+    temp = uniqueinds(Hm[1, x], is)
+    if length(temp) == 0
+      append!(right_links, [ir0])
+    elseif length(temp) == 1
+      append!(right_links, temp)
+    else
+      error("ITensor does not seem to be of the correct order")
+    end
+  end
+  return left_links, right_links
+end
+
+function convert_itensor_to_itensormatrix(tensor; kwargs...)
```

Comment:
Same comment as below, rename to something more specific to the application, like `local_mpo_blocks`. If we pass indices, for example as a tuple, we could dispatch on the length of the tuple, for example:
```julia
function local_mpo_blocks(t::ITensor, inds::NTuple{4,Index})
  # Implementation for bulk
end

function local_mpo_blocks(t::ITensor, inds::NTuple{3,Index})
  # Implementation for boundaries
end
```